Editor/ScriptAuditor.cs

```diff
@@ -189,6 +189,30 @@ private void AnalyzeMethodBody(ProjectReport projectReport, AssemblyDefinition a
                     {
                         continue;
                     }
+                    var type = (TypeReference) inst.Operand;
+                    if (type.IsGenericParameter)
+                    {
+                        bool isValueType = true; // assume it's value type
+                        var genericType = (GenericParameter) type;
+                        if (genericType.HasReferenceTypeConstraint)
+                        {
+                            isValueType = false;
+                        }
+                        else
+                        {
+                            foreach (var constraint in genericType.Constraints)
+                            {
+                                if (!constraint.IsValueType)
+                                    isValueType = false;
+                            }                            
+                        }
+
+                        if (!isValueType)
```

Comment:
I would not expect to see a *box* instruction if the compiler knows the generic parameter is indeed a reference type. Have you see any concrete scenario in which the compiler emitted a *box* for a generic parameter constrained to a *reference type* ?

Reply:
Yes:

`public class TestBoxing<T> where T : class
{
    private T myGenericObject;
    void Update()
    {
        if (myGenericObject != null) // boxing
        {
        }
    }
}`

which results in:

`    .method private hidebysig 
        instance void Update () cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 13 (0xd)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: ldfld !0 class TestBoxing`1<!T>::myGenericObject
        IL_0006: box !T
        IL_000b: pop
        IL_000c: ret
    } // end of method TestBoxing`1::Update`

(as per sharplab.io)

Reply:
Indeed. Weird to me but the specification for *box* is clear (I checked the it after finding 
https://stackoverflow.com/questions/3298040/c-sharp-generic-type-is-boxed)